### PR TITLE
Feature: Discord Webhook Alert Handler

### PR DESCRIPTION
**Pull request by mattnotmitt**
###### Required for all non-trivial PRs
- [x] Rebased/mergable
- [x] Tests pass
- [x] CHANGELOG.md updated
- [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

Added a [Discord Channel Webhook](https://discordapp.com/developers/docs/resources/webhook#execute-webhook) alert handler.

### TODO
- [x] Add service/discord/discordtest
- [x] Write documentation and contribute to https://github.com/influxdata/docs.influxdata.com
- [x] Comment code in pipeline/alert.go

### Screenshot of test notification in Discord client
img-2020-02-10-155610

### CODE REVIEW

File: pipeline/alert.go [hunk 1]
```diff
@@ -1523,6 +1528,94 @@ type SlackHandler struct {
...
+//      url = "https://discordapp.com/api/webhooks/xxxxxxxxxxxxxxxxxx/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
+//
+// In order to not post a message every alert interval
+// use AlertNode.StateChangesOnly so that only events
+// where the alert changed state are posted to the channel.
+//
+// Example:
+//    stream
+//         |alert()
+//             .slack()
```
Comment (goller) [inline pipeline/alert.go]: Would you be able to remove this slack() and the next comment?

File: pipeline/alert.go [hunk 2]
```diff
@@ -1523,6 +1528,94 @@ type SlackHandler struct {
...
+// Example:
+// stream
+//      |alert()
+//          .discord()
+//          .workspace('opencommunity')
+//
+// send alerts to the opencommunity workspace
+//
+// If the 'discord' section in the configuration has the option: global = true
+// then all alerts are sent to Slack without the need to explicitly state it
```
Comment (goller) [inline pipeline/alert.go]: (slack here)

File: server/server.go [hunk 3]
```diff
@@ -240,6 +241,9 @@ func New(c *Config, buildInfo BuildInfo, diagService *diagnostic.Service) (*Serv
 
 	// Append Alert integration services
 	s.appendAlertaService()
+	if err := s.appendDiscordService(); err != nil {
+		return nil, errors.Wrap(err, "slack service")
```
Comment (goller) [inline server/server.go]: (slack)

File: services/diagnostic/handlers.go [hunk 4]
```diff
@@ -598,6 +599,28 @@ func (h *SlackHandler) WithContext(ctx ...keyvalue.T) slack.Diagnostic {
 	}
 }
 
+// Slack Handler
```
Comment (goller) [inline services/diagnostic/handlers.go]: (slack)

File: services/discord/config.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+	URL string `toml:"url" override:"url"`
+	// Whether the timestamp is included in the embed
+	Timestamp bool `toml:"timestamp" override:"timestamp"`
+	// Username of webhook
+	Username string `toml:"username" override:"username"`
+	// Avatar URL
+	AvatarURL string `toml:"avatar-url" override:"avatar-url"`
+	// Embed Title
+	EmbedTitle string `toml:"embed-title" override:"embed-title"`
+	// Whether all alerts should automatically post to slack
```
Comment (goller) [inline services/discord/config.go]: (slack)

File: services/discord/service.go [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+func (s *Service) client(wid string) (*http.Client, error) {
+	w, err := s.workspace(wid)
+	if err != nil {
+		return &http.Client{}, err
+	}
+	return w.Client(), nil
+}
+
+func (s *Service) Update(newConfigs []interface{}) error {
+
```
Comment (goller) [inline services/discord/service.go]: formatting: extra space here

**Comment by mattnotmitt** [comment]
Ha, sorry, thought I'd caught all of those, pushing now!

**Comment by mattnotmitt** [comment]
I'm also happy to add the required stuff to Chronograf, I'd just need a pointer to what needs updating!

**Comment by desa** [comment]
@mattnotmitt is this ready for another review?

**Comment by mattnotmitt** [comment]
@desa very nearly, there's an issue I'm experiencing with the timestamp boolean - I posted about it in the kapacitor channel in the influxdata community slack.

**Comment by desa** [comment]
@mattnotmitt can you link me to the issue?

**Comment by mattnotmitt** [comment]
I've pinged you in the thread but here's the text:

I've been having some trouble with using a boolean in the HandlerConfig for
a custom alert handler. If you look at my pr
https://github.com/influxdata/kapacitor/pull/2287/files, the Timestamp is a
boolean field configurable in both in the `kapacitor.conf` file and when
the handler is called inline. Oddly, the value in the `kapacitor.conf` file
is ignored always, and the only way i can decide whether the timestamp
shows or not is if it is set in the tickscript. Could someone with
knowledge of the arcane internals of the kapacitor config loading give me a
hand? I'm sure i could figure it out myself but I'd appreciate some
guidance.

**Comment by mattnotmitt** [comment]
@desa Ready for review whenever you're available.

**Comment by desa** [comment]
@mattnotmitt were you able to take a look at my suggestion in slack?

**Comment by mattnotmitt** [comment]
Yeah I looked and replied. There's no discernable difference between it
being set to false and it just not being set. Decided to just take away the
handler config and keep it in the conf file.

**Comment by desa** [comment]
Typically people use pointers to differentiate between the two. I think that should work.

**Comment by mattnotmitt** [comment]
I don't really know how to implement that (my go skills aren't that great)
and the timestamp is so tiny that I find it unlikely that someone would
need to hide it. If you need this to merge I might be able to do it in a
few months but I need to knuckle down on some other work now, sorry.

**Comment by desa** [comment]
@mattnotmitt just reaching out with regards to

> I'm also happy to add the required stuff to Chronograf, I'd just need a pointer to what needs updating!

Is this something that you'd still be interested in doing? I'd be happy to offer an outline of the steps involved.

**Comment by mattnotmitt** [comment]
> @mattnotmitt just reaching out with regards to
> > ...
> Is this something that you'd still be interested in doing? I'd be happy to offer an outline of the steps involved.

Hi @desa, yeah I'd still be happy to work on that!

**Comment by desa** [comment]
@mattnotmitt Awesome. For 2.0 of InfluxDB, you're going to want to take a look at the code in https://github.com/influxdata/flux/blob/8a6ee0356c0837db0ace34cb7a2578be600874d3/stdlib/pagerduty/pagerduty.flux, https://github.com/influxdata/influxdb/tree/master/notification/endpoint, and  https://github.com/influxdata/influxdb/tree/master/notification/rule. This part of the 2.0 codebase is a bit rough with a fair amount of repeated logic.

For 1.x Chronograf, you'll want to start looking in https://github.com/influxdata/chronograf/blob/2dadcaf8dcb6a6733e027e1d9205f3688b81b62d/ui/src/types/kapacitor.ts and https://github.com/influxdata/chronograf/blob/9d8a49ba0ef8131cdce22d73718859f55f434db2/ui/src/kapacitor/components/handlers/index.js.

**Pull request merged**